### PR TITLE
Add FastAPI migration of licenses endpoints

### PR DESCRIPTION
**Pull request by davelopez**
Related to #10889
I decided to migrate the easiest API endpoints I found as a FastAPI exercise (and to break the ice) :)

Some comments:
- I decided to add a couple of [type annotated methods](https://github.com/galaxyproject/galaxy/compare/dev...davelopez:fastapi-migration-licenses?expand=1#diff-b0cd1bfe489edaa5d2f79359bf3718b115b1a36ac1cf104328e2021161f22a0aR79-R82) to the `LicensesManager` instead of using the existing ones so I can slightly change their behaviour while maintaining the previous controller intact.
- I assumed that the `/api/licenses/{id}` endpoint will always be invoked using an SPDX identifier as `Path` parameter and not an `url` (the manager supports both). If the URL must be supported the Path parameter should be changed to a Query.

TODO:
- [x] Add functional tests for the licenses API endpoints

### CODE REVIEW

File: lib/galaxy/managers/licenses.py [hunk 1]
```diff
@@ -1,10 +1,28 @@
 import json
 import logging
+from typing import List, Optional
```
Comment (mvdbeek) [inline lib/galaxy/managers/licenses.py]: We typically do multiple imports from the same module like so:
```suggestion
from typing import (
    List,
    Optional
)
```

File: lib/galaxy/managers/licenses.py [hunk 2]
```diff
@@ -1,10 +1,28 @@
...
 
 from pkg_resources import resource_string
+from pydantic import BaseModel, Field, HttpUrl
```
Comment (mvdbeek) [inline lib/galaxy/managers/licenses.py]:
```suggestion
from pydantic import (
    BaseModel,
    Field,
    HttpUrl
)
```

**Review by mvdbeek — COMMENTED** [review]
That's looking very nice!

File: lib/galaxy/managers/licenses.py (hunk 1 above)
Comment (davelopez) [reply lib/galaxy/managers/licenses.py]: Thanks Marius! Good to know :)

**Comment by jmchilton** [comment]
> I decided to add a couple of type annotated methods to the LicensesManager instead of using the existing ones so I can slightly change their behaviour while maintaining the previous controller intact.

This code hasn't been included in a Galaxy release yet - this is probably the best time to get everything perfect and synchronized. For instance, I would move that ``if license is None: raise ObjectNotFound()`` right into the manager.

**Review by davelopez — COMMENTED** [review]
Thanks for the quick review!

**Comment by davelopez** [comment]
I tried to fully document the API endpoints to serve as an example.
Also, I took the liberty to add documentation for the API tags [here](https://github.com/galaxyproject/galaxy/pull/11056/commits/d9ff4b1b706b5634c2de987150f6bdc7107f5fff) as explained in the [FastAPI documentation](https://fastapi.tiangolo.com/tutorial/metadata/#metadata-for-tags).

File: lib/galaxy/webapps/galaxy/fast_app.py [hunk 3]
```diff
@@ -5,7 +5,24 @@
...
+        "name": "licenses",
+        "description": "Operations with [SPDX licenses](https://spdx.org/licenses/).",
+    },
+]
+
+
+def add_open_api_tags_metadata(app: FastAPI) -> None:
+    """Adds OpenAPI documentation to the APIRouter tags"""
+    if app:
+        app.openapi_tags = api_tags_metadata
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/fast_app.py]: Could we just add this on line 52 as the docs do (`FastAPI(openapi_tags=api_tags_metadata)`)? I'm not sure setting the attribute on the instance is really the public API we should be using?

Comment (davelopez) [reply lib/galaxy/webapps/galaxy/fast_app.py]: Yes, that was my first approach, but then I noticed that when launching from `/scripts/fapi.py` the app was already instanced in `fapi.py` [line 12](https://github.com/galaxyproject/galaxy/blob/dev/scripts/fapi.py#L12). I am not sure if this script will be just temporal and we assume that the only creation point will be on line 52.

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/fast_app.py]: I think we can do this:
```
diff --git a/scripts/fapi.py b/scripts/fapi.py
index c24d54e9bf..1d94458bce 100644
--- a/scripts/fapi.py
+++ b/scripts/fapi.py
@@ -1,5 +1,4 @@
 import uvicorn
-from fastapi import FastAPI
 from galaxy_main import (
     GalaxyConfigBuilder,
     main,
@@ -9,16 +8,13 @@ from galaxy.util.properties import load_app_properties
 from galaxy.webapps.galaxy.buildapp import app_factory
 from galaxy.webapps.galaxy.fast_app import initialize_fast_app
 
-app = FastAPI()
-
 
 def wsgiloop(args, log):
     config_builder = GalaxyConfigBuilder(args)
     kwds = config_builder.app_kwds()
     kwds = load_app_properties(**kwds)
     gx = app_factory(global_conf=config_builder.global_conf(), **kwds)
-    initialize_fast_app(gx, app=app)
-    uvicorn.run(app)
+    uvicorn.run(initialize_fast_app(gx))
 
 
 if __name__ == '__main__':

```
I only created the app object in the global namespace so we could use gunicorn, but that doesn't work yet with the way we the configuration is read.

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/fast_app.py]: I'll open a separate PR for that so we can keep this PR as a clean example for how to do migrations.

File: lib/galaxy_test/api/test_licenses.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        self._assert_status_code_is(response, 200)
+
+    def test_get_license(self):
+        licenseId = "Apache-2.0"
+        response = self._get(f"licenses/{licenseId}")
+        self._assert_status_code_is(response, 200)
+        self._assert_matches_license_id(licenseId, response.json())
+
+    def test_404_on_unknown_license(self):
+        licenseId = "uknown"
```
Comment (mvdbeek) [inline lib/galaxy_test/api/test_licenses.py]:
```suggestion
        licenseId = "unknown"
```

File: lib/galaxy_test/api/test_licenses.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        response = self._get(f"licenses/{licenseId}")
+        self._assert_status_code_is(response, 404)
```
Comment (mvdbeek) [inline lib/galaxy_test/api/test_licenses.py:19]: Haha, so you found a bug in the old API route, which returns a 200 here. Can you use `get_license_by_id` in the old endpoint too ?

File: lib/galaxy/webapps/galaxy/fast_app.py (hunk 3 above)
Comment (davelopez) [reply lib/galaxy/webapps/galaxy/fast_app.py]: Cool! I'll remove the `add_open_api_tags_metadata` then :)

File: lib/galaxy_test/api/test_licenses.py (hunk 5 above)
Comment (davelopez) [reply lib/galaxy_test/api/test_licenses.py:19]: Sure! :)

**Comment by davelopez** [comment]
Did I break the `RolesApiTestCase::test_create_only_admin` test somehow? :flushed:

**Comment by mvdbeek** [comment]
I've seen this fail before in other PRs, just one of those flaky tests we'll try fixing on Monday.

**Review by mvdbeek — APPROVED** [review]

File: lib/galaxy/webapps/galaxy/fast_app.py [hunk 6]
```diff
@@ -30,6 +41,7 @@ async def message_exception_middleware(request: Request, exc: MessageException)
         )
 
 
+
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/fast_app.py]:
```suggestion
```
sorry about that, I should stop using the github interface to do these things.

**Comment by davelopez** [comment]
Thanks for resolving the conflict! :)

**Comment by mvdbeek** [comment]
Thank you @davelopez!

**Pull request merged**